src/ray/gcs/store_client/in_memory_store_client.cc

```diff
@@ -0,0 +1,173 @@
+  absl::MutexLock lock(&mutex_);
+  records_[key] = data;
+  index_keys_[index_key].push_back(key);
+}
+
+boost::optional<std::string> InMemoryTable::Get(const std::string &key) {
+  absl::MutexLock lock(&mutex_);
+  if (records_.count(key)) {
+    return records_[key];
+  }
+  return boost::none;
+}
+
+std::unordered_map<std::string, std::string> InMemoryTable::GetAll() {
+  absl::MutexLock lock(&mutex_);
+  return records_;
+}
+
+void InMemoryTable::Delete(const std::string &key) {
+  absl::MutexLock lock(&mutex_);
+  records_.erase(key);
+}
+
+void InMemoryTable::DeleteByIndex(const std::string &index_key) {
+  absl::MutexLock lock(&mutex_);
+  auto delete_keys = index_keys_[index_key];
+  for (auto &delete_key : delete_keys) {
+    records_.erase(delete_key);
+  }
+  index_keys_.erase(index_key);
+}
+
+template <typename Key, typename Data, typename IndexKey>
+Status InMemoryStoreClient<Key, Data, IndexKey>::AsyncPut(
+    const std::string &table_name, const Key &key, const Data &data,
+    const StatusCallback &callback) {
+  auto table = GetOrCreateTable(table_name);
+  io_service_.post([table, key, data, callback]() {
+    table->Put(key.Binary(), data.SerializeAsString());
+    callback(Status::OK());
+  });
+  return Status::OK();
+}
+
+template <typename Key, typename Data, typename IndexKey>
+Status InMemoryStoreClient<Key, Data, IndexKey>::AsyncPutWithIndex(
+    const std::string &table_name, const Key &key, const IndexKey &index_key,
+    const Data &data, const StatusCallback &callback) {
+  auto table = GetOrCreateTable(table_name);
+  io_service_.post([table, key, index_key, data, callback]() {
+    table->PutWithIndex(key.Binary(), index_key.Binary(), data.SerializeAsString());
+    callback(Status::OK());
+  });
+  return Status::OK();
+}
+
+template <typename Key, typename Data, typename IndexKey>
+Status InMemoryStoreClient<Key, Data, IndexKey>::AsyncGet(
+    const std::string &table_name, const Key &key,
+    const OptionalItemCallback<Data> &callback) {
+  auto table = GetOrCreateTable(table_name);
+  io_service_.post([table, key, callback]() {
+    auto value = table->Get(key.Binary());
+    if (value) {
+      Data data;
+      RAY_CHECK(data.ParseFromString(*value));
+      callback(Status::OK(), data);
+    } else {
+      callback(Status::OK(), boost::none);
+    }
+  });
+  return Status::OK();
+}
+
+template <typename Key, typename Data, typename IndexKey>
+Status InMemoryStoreClient<Key, Data, IndexKey>::AsyncGetAll(
+    const std::string &table_name,
+    const SegmentedCallback<std::pair<Key, Data>> &callback) {
+  auto table = GetOrCreateTable(table_name);
+  io_service_.post([table, callback]() {
+    auto records = table->GetAll();
+    std::vector<std::pair<Key, Data>> result;
+    for (auto &record : records) {
+      Data data;
+      RAY_CHECK(data.ParseFromString(record.second));
+      result.push_back(std::make_pair(Key::FromBinary(record.first), data));
```

Comment:
use `emplace_back`, and std::move(data).

Reply:
fixed